### DISCUSSION TITLE
Insert date/time into PS1 prompt

### DISCUSSION THREAD

**Discussion by haydenflinner** [zimfw/zimfw#461] [Q&A] [answered]
Hi, new zimfw user here. Switched off of oh-my-zsh because I found it sluggish when spending a lot of time on a slow NFS. There's one thing I'm missing.

I am using the default theme, and I'd like to add the output of $(date) to the PS1 prompt, so that when I come back to an old tmux pane I can see how long ago cmds were run / how long I spent iterating on a particular project. Unfortunately this isn't as intuitive for me as default zsh, since it looks like in the [theme definition file](https://github.com/zimfw/asciiship/blob/master/asciiship.zsh-theme) we're defining PS1 to be some template which gets expanded later, and when I try inserting an arbitrary cmd there `$(date)`, it doesn't get expanded and rendered properly. I am guessing this is an obvious tweak for someone who knows more zsh, so didn't spend too much time tinkering. Thanks for the framework and thanks in advance for any pointers you can spare!

**Comment by ericbn** [comment, marked as answer]
Hi @haydenflinner and welcome to Zim!  :- )

The definition of the left prompt itself is set in the `PS1` variable:
```zsh
PS1='
%(2L.%B%F{yellow}(%L)%f%b .)%(!.%B%F{red}%n%f%b in .${SSH_TTY:+"%B%F{yellow}%n%f%b in "})${SSH_TTY:+"%B%F{green}%m%f%b in "}%B%F{cyan}%~%f%b${(e)git_info[prompt]}${VIRTUAL_ENV:+" via %B%F{yellow}${VIRTUAL_ENV:t}%b%f"}${duration_info}
%B%(1j.%F{blue}*%f .)%(?.%F{green}.%F{red}%? )%#%f%b '
```
Let me break that up for you:
* Starts with a line break, which adds a blank line to the prompt.
* `%(2L.%B%F{yellow}(%L)%f%b .)` outputs:
  * If the `SHLVL` parameter is at least `2`: `%B%F{yellow}(%L)%f%b `, which is `(`, the current value of `$SHLVL`, `)` in bold yellow.
  * Else nothing.
* `%(!.%B%F{red}%n%f%b in .${SSH_TTY:+"%B%F{yellow}%n%f%b in "})` outputs:
  * If the shell is running with privileges: `%B%F{red}%n%f%b in `, which is `$USERNAME` in bold red followed by <code>&nbsp;in&nbsp;</code>.
  * Else: `${SSH_TTY:+"%B%F{yellow}%n%f%b in "}`, which is `$USERNAME` in bold yellow followed by <code>&nbsp;in&nbsp;</code>, but only if `SSH_TTY` is non-null.
* `${SSH_TTY:+"%B%F{green}%m%f%b in "}`, which is the hostname up to the first `.` in bold green followed by <code>&nbsp;in&nbsp;</code>, but only if `SSH_TTY` is non-null.
* `%B%F{cyan}%~%f%b`, which outputs the current working directory in bold cyan.
* `${(e)git_info[prompt]}`
* `${VIRTUAL_ENV:+" via %B%F{yellow}${VIRTUAL_ENV:t}%b%f"}`, which is <code>&nbsp;via&nbsp;</code> followed by the final path component of `VIRTUAL_ENV` in bold yellow, but only if `VIRTUAL_ENV` is non-null.
* `${duration_info}`
* Line break.
* `%B%(1j.%F{blue}*%f .)%(?.%F{green}.%F{red}%? )%#%f%b` which is all the following in bold:
  * `%(1j.%F{blue}*%f .)`, which:
    * If the number of jobs is at least `1`: outputs `*` in blue.
    * Else nothing.
  * `%(?.%F{green}.%F{red}%? )`, which:
    * If the exit status of the last command was `0`: sets the foreground color to green.
    * Else sets the foreground color to red, and outputs the exit status number.
  * `%#`, which is a `#` if the shell is running with privileges, a `%` if not.
  * `%f`, which resets the foreground color.
* A space.

You can read about all the `%x` prompt expansion sequences here: https://zsh.sourceforge.io/Doc/Release/Prompt-Expansion.html. And about the parameter expansion like `${name:+word}`, here: https://zsh.sourceforge.io/Doc/Release/Expansion.html#Parameter-Expansion.

Note that there are prompt expansion sequences to output the date and time, so no need to start a new subshell with `$(date)`. For example, you could add <code>&nbsp;at %B%F{yellow}%D %T%f%b$</code> before `${duration_info}`.

If you're using degit, local changes are overwritten on updates, so be careful with that. In any case, you can create a fork of the [asciiship](https://github.com/zimfw/asciiship) repository under your username with your changes there, and then replace the zmodule call by `zmodule haydenflinner/asciiship`, which is nicer for you to carry around if you want install it in different machines...

**Reply by haydenflinner** [reply]
Thank you so much for the in depth explanation! And for the link to Prompt Expansion, I had assumed it was the framework doing the expansion and not zsh itself!
